### ISSUE TITLE
Kubelet requires restart after granting AKS SP access to ACR resource

### ISSUE THREAD

**Issue by neilpeterson** [Azure/AKS#442]
It seems that after granting the AKS SP access to an ACR resource ([docs](https://docs.microsoft.com/en-us/azure/container-registry/container-registry-auth-aks?toc=%2fazure%2faks%2ftoc.json#grant-aks-access-to-acr)), that the kubelet on each AKS nodes requires a restart before container images can be pulled.

Is there any way to avoid this?

Related issue on the doc repo - https://github.com/MicrosoftDocs/azure-docs/issues/10118

**Comment by andyzhangx** [comment]
I also found this, will take a look:
```
Events:
  Type     Reason                 Age              From                               Message
  ----     ------                 ----             ----                               -------
  Warning  FailedScheduling       8m (x3 over 8m)  default-scheduler                  0/1 nodes are available: 1 Insufficient cpu.
  Normal   Scheduled              8m               default-scheduler                  Successfully assigned nginx-server-776564f79c-zhtjk to aks-nodepool1-20881069-0
  Normal   SuccessfulMountVolume  8m               kubelet, aks-nodepool1-20881069-0  MountVolume.SetUp succeeded for volume "default-token-4t7tk"
  Normal   SuccessfulMountVolume  8m               kubelet, aks-nodepool1-20881069-0  MountVolume.SetUp succeeded for volume "pvc-5c1f0521-739f-11e8-9b69-0a58ac1f09c2"
  Warning  Failed                 8m (x5 over 8m)  kubelet, aks-nodepool1-20881069-0  Error: ImagePullBackOff
  Normal   BackOff                8m (x5 over 8m)  kubelet, aks-nodepool1-20881069-0  Back-off pulling image "andyacr.azurecr.io/nginx-server:1.0.0"
  Warning  Failed                 8m (x2 over 8m)  kubelet, aks-nodepool1-20881069-0  Error: ErrImagePull
  Warning  Failed                 8m (x2 over 8m)  kubelet, aks-nodepool1-20881069-0  Failed to pull image "andyacr.azurecr.io/nginx-server:1.0.0": rpc error: code = Unknown desc = Error response from daemon: Get https://andyacr.azurecr.io/v2/nginx-server/manifests/1.0.0: unauthorized: authentication required
  Normal   Pulling                8m (x2 over 8m)  kubelet, aks-nodepool1-20881069-0  pulling image "andyacr.azurecr.io/nginx-server:1.0.0"
  Normal   SandboxChanged         8m (x7 over 8m)  kubelet, aks-nodepool1-20881069-0  Pod sandbox changed, it will be killed and re-created.
  Warning  NetworkNotReady        3m               kubelet, aks-nodepool1-20881069-0  network is not ready: [runtime network not ready: NetworkReady=false reason:NetworkPluginNotReady message:docker: network plugin is not ready: Kubenet does not have netConfig. This is most likely due to lack of PodCIDR]
  Normal   SuccessfulMountVolume  3m               kubelet, aks-nodepool1-20881069-0  MountVolume.SetUp succeeded for volume "pvc-5c1f0521-739f-11e8-9b69-0a58ac1f09c2"
  Normal   SuccessfulMountVolume  3m               kubelet, aks-nodepool1-20881069-0  MountVolume.SetUp succeeded for volume "default-token-4t7tk"
  Normal   Pulling                3m               kubelet, aks-nodepool1-20881069-0  pulling image "andyacr.azurecr.io/nginx-server:1.0.0"
  Normal   Pulled                 3m               kubelet, aks-nodepool1-20881069-0  Successfully pulled image "andyacr.azurecr.io/nginx-server:1.0.0"
  Normal   Created                3m               kubelet, aks-nodepool1-20881069-0  Created container
  Normal   Started                3m               kubelet, aks-nodepool1-20881069-0  Started container
```

**Comment by andyzhangx** [comment]
@neilpeterson Unfortunately this is a bug in all k8s version, restart kubelet should be the only workaround now. FYI @khenidak I have filed a bug in k8s upstream, will fix this issue soon.

**Comment by neilpeterson** [comment]
@andyzhangx thanks for following up on this, appreciate it.

**Comment by andyzhangx** [comment]
paste another workaround here:
```
1)	create sp first
2)	grant sp access to acr
3)	create AKS cluster by sp:
az aks create --resource-group $RESOURCE_GROUP_NAME --name $CLUSTER_NAME --node-count 1 --generate-ssh-keys --service-principal $APP_ID --client-secret $CLIENT_SECRET 
```

**Comment by andyzhangx** [comment]
By a deeper investigation, I found the acrProvider will refrech every 12 min, so if this error exists, just wait at most 12min, this issue would disappear.
```
Jul 09 11:42:31 aks-nodepool1-27069170-0 kubelet[8452]: I0709 11:42:31.893471    8452 provider.go:119] Refreshing cache for provider: *credentialprovider.defaultDockerConfigProvider
Jul 09 11:42:31 aks-nodepool1-27069170-0 kubelet[8452]: I0709 11:42:31.893875    8452 provider.go:119] Refreshing cache for provider: *azure.acrProvider
Jul 09 11:53:54 aks-nodepool1-27069170-0 kubelet[8452]: I0709 11:53:54.171097    8452 provider.go:119] Refreshing cache for provider: *credentialprovider.defaultDockerConfigProvider
Jul 09 11:53:54 aks-nodepool1-27069170-0 kubelet[8452]: I0709 11:53:54.171530    8452 provider.go:119] Refreshing cache for provider: *azure.acrProvider
```

**Comment by brusMX** [comment]
I know this is not a solution to this problem, but I just wanted to share a workaround for anyone that wants to provide access to an ACR created after the k8s cluster provision.  
You can just create a secret (if you have the kubectl access). This effectively doesn't solve this issue, but might unblock you if you need to add an ACR:

```bash
kubectl create secret docker-registry acr-auth --docker-server ${registryName}.azurecr.io --docker-username $CLIENT_ID --docker-password $SP_PASSWD --docker-email email@k8scluster.com
```
Source: https://github.com/MicrosoftDocs/azure-docs/blob/master/articles/container-registry/container-registry-auth-aks.md#access-with-kubernetes-secret

**Comment by andyzhangx** [comment]
@brusMX yes, that's a common way to store registry secret. I already got a fix for this: https://github.com/kubernetes/kubernetes/pull/66429

Here is the root cause:
after granting sp access to azure ACR , pull image from ACR would fail, and **after wait about 15-30min, pull image would succeed**. Root cause is that `servicePrincipalToken` needs to be refreshed when doing `registryClient.List`, otherwise it will always return empty registry list.

**Comment by andyzhangx** [comment]
Update:
The [fix](https://github.com/kubernetes/kubernetes/pull/66429) has already been merged in upstream, and also cherry-picked to v1.9 - v1.11

**Comment by brusMX** [comment]
Thank you so much @andyzhangx !

**Comment by alastairs** [comment]
I just deployed a new AKS cluster running Kubernetes 1.11.2, and the issue is still present. Is there any update available on this issue?

**Comment by andyzhangx** [comment]
@alastairs what's your error msg? And could you check whether your service principal has access to acr by:
```
az login --service-principal -u <aadClientId> -p <aadClientSecret> -t <tenantId>
az acr list
```

**Comment by alastairs** [comment]
@andyzhangx I had a similar error message to the one described in Azure/acs-engine#3654 above. I can't now try the `az login` test you mentioned as I've lost the password to the service principal 😞 I followed the instructions on [authenticating with a Kubernetes secret](https://docs.microsoft.com/en-us/azure/container-registry/container-registry-auth-aks#access-with-kubernetes-secret), but this didn't resolve the issue, even after rebooting the AKS agent nodes (to restart kubelet). 

Eventually I followed the instructions on [Granting AKS access to ACR](https://docs.microsoft.com/en-us/azure/container-registry/container-registry-auth-aks#grant-aks-access-to-acr) instead, and this did resolve the issue. 

Perhaps this is a separate issue, in which case I apologise for the distraction here. I was surprised to see this was still open given [your comment](https://github.com/Azure/AKS/issues/442#issuecomment-407994578) that a fix had been merged upstream and into 1.11.2.

**Comment by andyzhangx** [comment]
@alastairs Thanks for check, your issue is due to "sp has no access to ACR", it's a config issue.

@neilpeterson This issue has been fixed in following versions (or later), would you close this issue? thanks.
```
| k8s version | fixed version |
| ---- | ---- |
| v1.8 | no fix |
| v1.9 | 1.9.11 |
| v1.10 | 1.10.7 |
| v1.11 | 1.11.2 |
| v1.12 | 1.12.0 |
```

**Comment by andyzhangx** [comment]
@HeyHugo Your error is `lookup <my-acr-host> 168.63.129.16:53: no such host`, could you run `docker pull <my-acr-host>/app:latest` to make sure you have the right ACR address.
And after you create a service principal, you need to grant ACR read permission to that sp, if `az acr list` shows an empty list, it means sp does not have ACR permission.
Restart kubelet won't work in your case.

**Comment by andyzhangx** [comment]
@HeyHugo pls make sure you have do the following:
```
# Create role assignment
az role assignment create --assignee $CLIENT_ID --role acrpull --scope $ACR_ID
```
And after you create a service principal, you need to grant ACR read permission to that sp, if az acr list shows an empty list, it means sp does not have ACR permission.

**Comment by HeyHugo** [comment]
Ok thanks @andyzhangx.  So I had to add "Reader" role assignment throught the portal web UI. Now I can see the container registry when I log in as the service principal.

However my pods still get the same error when trying to pull images.

**Comment by andyzhangx** [comment]
@HeyHugo could you paste your `kubectl describe po POD-NAME` error? In your original error, it's `lookup <my-acr-host> 168.63.129.16:53: no such host`, looks like your node could not resove your ACR address.

**Comment by HeyHugo** [comment]
Ugh 🤦‍♂️ sorry there was a typo in the hostname in my helm templates. Thanks for all the help @andyzhangx now it all seems to work.

**Comment by andyzhangx** [comment]
great, I will close this issue, just let me know if you have any similar issue.
